src/wallet/asyncrpcoperation_sendmany.cpp

```diff
@@ -191,791 +149,508 @@ void AsyncRPCOperation_sendmany::main() {
+        assert(pwalletMain->GetSproutSpendingKey(t.address, sk));
 
-            jsInputValue += noteFunds;
+        builder_.AddSproutInput(sk, t.note, vSproutWitnesses[i].value());
 
-            int wtxHeight = -1;
-            int wtxDepth = -1;
-            {
-                LOCK2(cs_main, pwalletMain->cs_wallet);
-                const CWalletTx& wtx = pwalletMain->mapWallet[jso.hash];
-                // Zero-confirmation notes belong to transactions which have not yet been mined
-                if (mapBlockIndex.find(wtx.hashBlock) == mapBlockIndex.end()) {
-                    throw JSONRPCError(RPC_WALLET_ERROR, strprintf("mapBlockIndex does not contain block hash %s", wtx.hashBlock.ToString()));
-                }
-                wtxHeight = mapBlockIndex[wtx.hashBlock]->nHeight;
-                wtxDepth = wtx.GetDepthInMainChain();
-            }
-            LogPrint("zrpcunsafe", "%s: spending note (txid=%s, vJoinSplit=%d, jsoutindex=%d, amount=%s, height=%d, confirmations=%d)\n",
-                    getId(),
-                    jso.hash.ToString().substr(0, 10),
-                    jso.js,
-                    int(jso.n), // uint8_t
-                    FormatMoney(noteFunds),
-                    wtxHeight,
-                    wtxDepth
-                    );
+        sum += t.note.value();
+        if (sum >= txValues.targetAmount) {
+            break;
         }
+    }
 
-        // Add history of previous commitments to witness
-        if (vInputNotes.size() > 0) {
-
-            if (vInputWitnesses.size()==0) {
-                throw JSONRPCError(RPC_WALLET_ERROR, "Could not find witness for note commitment");
-            }
-
-            for (auto & optionalWitness : vInputWitnesses) {
-                if (!optionalWitness) {
-                    throw JSONRPCError(RPC_WALLET_ERROR, "Witness for note commitment is null");
-                }
-                SproutWitness w = *optionalWitness; // could use .value();
-                if (jsChange > 0) {
-                    for (const uint256& commitment : previousCommitments) {
-                        w.append(commitment);
-                    }
-                    if (jsAnchor != w.root()) {
-                        throw JSONRPCError(RPC_WALLET_ERROR, "Witness for spendable note does not have same anchor as change input");
-                    }
-                }
-                witnesses.push_back(w);
-            }
-
-            // The jsAnchor is null if this JoinSplit is at the start of a new chain
-            if (jsAnchor.IsNull()) {
-                jsAnchor = inputAnchor;
-            }
+    // Build the transaction
+    auto buildResult = builder_.Build();
+    auto tx = buildResult.GetTxOrThrow();
 
-            // Add spendable notes as inputs
-            std::copy(vInputNotes.begin(), vInputNotes.end(), std::back_inserter(info.notes));
-        }
+    UniValue sendResult = SendTransaction(tx, keyChange, testmode);
```

Comment:
I would split out a part that creates and returns the transaction without sending it, so that it can more easily be tested independently. (I know this part still takes global locks and reads from the blockchain.)

Reply:
I think this is a good idea, but I think the right thing to do here is factor out a wallet-level transaction builder that then gets reused in z_shieldcoinbase, z_mergetoaddress etc. that implements the wallet rules. The current implementation retains the behavior that sending to the network is part of the async operation.